components/logins/android/src/main/java/mozilla/appservices/logins/DatabaseLoginsStorage.kt

```diff
@@ -111,60 +111,40 @@ class DatabaseLoginsStorage(private val dbPath: String) : AutoCloseable {
     }
 
     @Throws(LoginsStorageException::class)
-    fun list(): List<Login> {
+    fun decryptAndFixupLogin(encKey: String, login: Login): LoginPayload {
         return readQueryCounters.measure {
-            checkUnlocked().list()
+            checkUnlocked().decryptAndFixupLogin(encKey, login)
         }
     }
 
     @Throws(LoginsStorageException::class)
-    fun getByBaseDomain(baseDomain: String): List<Login> {
+    fun list(): List<Login> {
         return readQueryCounters.measure {
-            checkUnlocked().getByBaseDomain(baseDomain)
+            checkUnlocked().list()
         }
     }
 
     @Throws(LoginsStorageException::class)
-    fun add(login: Login): String {
-        return writeQueryCounters.measure {
-            checkUnlocked().add(login)
+    fun getByBaseDomain(baseDomain: String): List<Login> {
+        return readQueryCounters.measure {
+            checkUnlocked().getByBaseDomain(baseDomain)
         }
     }
 
     @Throws(LoginsStorageException::class)
-    fun importLogins(logins: List<Login>): String {
+    fun addOrUpdate(encKey: String, login: Login): String {
```

Comment:
`addOrUpdate()` is the biggest question for me.  I'm going to copy/rephrase my thoughts from the google doc here.  BTW, thanks for making this PR, it's nice to have the design more in the open.

Based on the way the API is currently shaped, it seems like you should use the API like this:

 - When the user clicks on the username/password field, call getByBaseDomain() and present some of them to the user
 - If the user clicks on one, then autofil the form and store that `Login` somewhere
 - If the user changes the username/password field values, then update the username/password fields for that stored `Login`
 - If the user submits the form and clicks "Save this login", send the stored `Login` to `addOrUpdate()`.  `addOrUpdate()` will then update that `Login` since it already existed in the DB.

However, depending on what the user entered into the username field, we aren't going to update that login.  If they enter a new username then we should insert a new record.  If they enter an existing username, then we should update a different record. 

I think the issue is that `addOrUpdate()` seems like it would work like a `save()` method in an ORM: it will do an `INSERT` if the guid is not set and an `UPDATE ... WHERE guid={guid}` if the guid it set.  But actually, the GUID is more like a hint.  It only matters as a tiebreaker when there are duplicate records with the same username.   (At least this how I'm reading  [shouldUpdateOrCreateAsync()](https://github.com/mozilla-mobile/android-components/blob/d7b1b62da50cf224fce1eec7a163d858fb1a2334/components/service/sync-logins/src/main/java/mozilla/components/service/sync/logins/DefaultLoginValidationDelegate.kt#L29) and [ensureValidLogin()](https://github.com/mozilla-mobile/android-components/blob/d7b1b62da50cf224fce1eec7a163d858fb1a2334/components/service/sync-logins/src/main/java/mozilla/components/service/sync/logins/DefaultLoginValidationDelegate.kt#L74))

So I think the signature should change to better indicate what's happening:
  - Define a `LoginData` dictionary, which is the same as `Login`, but without the GUID field
  - Make the signature `addOrUpdate(encKey: String, login: LoginData, selectedGuid: String?)`

Also, on the consumer side we need to store the `Login` the user selects, but we shouldn't be mutating it.  Instead, we should mutate a separate `LoginData` value.  This is because we don't really know which `Login` is going to be updated until the entire process is complete.

Reply:
IIUC the login returned from ` shouldUpdateOrCreateAsync` is merged [here](https://github.com/mozilla-mobile/android-components/blob/bfaf514793dd20f9f12d3f551f9ff3e470f52727/components/service/sync-logins/src/main/java/mozilla/components/service/sync/logins/DefaultLoginValidationDelegate.kt#L76) with the selected login (the one passed into the [onLoginSave function](https://github.com/mozilla-mobile/android-components/blob/bfaf514793dd20f9f12d3f551f9ff3e470f52727/components/service/sync-logins/src/main/java/mozilla/components/service/sync/logins/GeckoLoginStorageDelegate.kt#L68) where `shouldUpdateOrCreateAsync` is called via `ensureValidAsync`) and the resulting merged login is then passed to the [update function](https://github.com/mozilla-mobile/android-components/blob/bfaf514793dd20f9f12d3f551f9ff3e470f52727/components/service/sync-logins/src/main/java/mozilla/components/service/sync/logins/GeckoLoginStorageDelegate.kt#L73). So while I agree the login guid is used as more of a tool to assess the precedence of the potential dupes in the `shouldUpdateOrCreateAsync` function, the login guid dictates which login is updated thereafter making it more consequential. 

Outside of the potential delays the necessary product and/or external team dependencies required to make this update, I question whether this change should be included in an effort that should be invisible to the user. Having said that, I think it's worth capturing this in a ticket for future consideration.

Reply:
My thinking was that we were already going to have to change the consumer code, so it wouldn't be much extra effort.  But if that's not true then I definitely agree that we should try to limit the scope.  In that case, what about adding some comments to the `addOrUpdate()` that describe how it's used and what the GUID field is doing?